src/DynamoRevit/Models/RevitDynamoModel.cs

```diff
@@ -382,37 +381,36 @@ private void SetupPython()
         {
             if (setupPython) return;
 
-            IronPythonEvaluator.OutputMarshaler.RegisterMarshaler(
-                (Element element) => element.ToDSType(true));
-
-            // Turn off element binding during iron python script execution
-            IronPythonEvaluator.EvaluationBegin +=
-                (a, b, c, d, e) => ElementBinder.IsEnabled = false;
-            IronPythonEvaluator.EvaluationEnd += (a, b, c, d, e) => ElementBinder.IsEnabled = true;
-
             var marshaler = new DataMarshaler();
             marshaler.RegisterMarshaler(
                 (Revit.Elements.Element element) => element.InternalElement);
             marshaler.RegisterMarshaler((Category element) => element.InternalCategory);
             Func<object, object> unwrap = marshaler.Marshal;
-            // register UnwrapElement method in ironpython
-            IronPythonEvaluator.EvaluationBegin += (a, b, scope, d, e) =>
-            {
-                scope.SetVariable("UnwrapElement", unwrap);
-            };
-
-            CPythonEvaluator.OutputMarshaler.RegisterMarshaler(
-                (Element element) => element.ToDSType(true));
 
-            // Turn off element binding during cpython script execution
-            CPythonEvaluator.EvaluationBegin +=
-                (a, b, c, d) => ElementBinder.IsEnabled = false;
-            CPythonEvaluator.EvaluationEnd += (a, b, c, d) => ElementBinder.IsEnabled = true;
 
-            // register UnwrapElement method in cpython
-            CPythonEvaluator.EvaluationBegin += (a, scope, c, d) =>
+            Action<PythonServices.PythonEngine> setupPythonEngine = (PythonServices.PythonEngine engine) =>
             {
-                scope.Set("UnwrapElement", unwrap);
+                (engine.OutputDataMarshaler as DataMarshaler).RegisterMarshaler((Element element) => element.ToDSType(true));
+
+                // Turn off element binding during python script execution
+                engine.EvaluationStarted += (a, b, c) => ElementBinder.IsEnabled = false;
+                // register UnwrapElement method
+                engine.EvaluationStarted += (a, b, scopeSet) => scopeSet("UnwrapElement", unwrap);
+                // Turn on element binding after python script execution
+                engine.EvaluationFinished += (a, b, c, d) => ElementBinder.IsEnabled = true;
+            };
+            // Setup engines for all existing python engines
+            PythonServices.PythonEngineManager.Instance.AvailableEngines.ToList().ForEach(engine => setupPythonEngine(engine));
+            // Setup engines for any python engines that might be registered later on
+            PythonServices.PythonEngineManager.Instance.AvailableEngines.CollectionChanged += (object sender, NotifyCollectionChangedEventArgs e) =>
```

Comment:
Do we need to unsubscribe these somewhere?

Reply:
It did not unsubscribe them before...I did not want to change around too much
Will take a look if I can unbsub them easily